drivers/gpio/gpio_npcx.c

```diff
@@ -0,0 +1,294 @@
+{
+	struct gpio_reg *inst = HAL_INSTANCE(dev);
+
+	/* Get raw bits of GPIO input registers */
+	*value = inst->PDIN;
+
+	return 0;
+}
+
+static int gpio_npcx_port_set_masked_raw(struct device *dev,
+					  gpio_port_pins_t mask,
+					  gpio_port_value_t value)
+{
+	struct gpio_reg *inst = HAL_INSTANCE(dev);
+	uint8_t out = inst->PDOUT;
+
+	inst->PDOUT = ((out & ~mask) | (value & mask));
+
+	return 0;
+}
+
+static int gpio_npcx_port_set_bits_raw(struct device *dev,
+					gpio_port_value_t mask)
+{
+	struct gpio_reg *inst = HAL_INSTANCE(dev);
+
+	/* Set raw bits of GPIO output registers */
+	inst->PDOUT |= mask;
+
+	return 0;
+}
+
+static int gpio_npcx_port_clear_bits_raw(struct device *dev,
+						gpio_port_value_t mask)
+{
+	struct gpio_reg *inst = HAL_INSTANCE(dev);
+
+	/* Clear raw bits of GPIO output registers */
+	inst->PDOUT &= ~mask;
+
+	return 0;
+}
+
+static int gpio_npcx_port_toggle_bits(struct device *dev,
+						gpio_port_value_t mask)
+{
+	struct gpio_reg *inst = HAL_INSTANCE(dev);
+
+	/* Toggle raw bits of GPIO output registers */
+	inst->PDOUT ^= mask;
+
+	return 0;
+}
+
+static int gpio_npcx_pin_interrupt_configure(struct device *dev,
+		gpio_pin_t pin, enum gpio_int_mode mode,
+		enum gpio_int_trig trig)
+{
+	const struct gpio_npcx_config *config = DRV_CONFIG(dev);
+	enum miwu_int_mode miwu_mode = NPCX_MIWU_MODE_DISABLED;
+	enum miwu_int_trig miwu_trig = NPCX_MIWU_TRIG_NONE;
+
+	if (config->wui_maps[pin].table == NPCX_MIWU_TABLE_NONE) {
+		LOG_ERR("Cannot configure GPIO(%x, %d)", config->port, pin);
+		return -EINVAL;
+	}
+
+	LOG_DBG("pin_int_conf (%d, %d) match (%d, %d, %d)!!!",
+			config->port, pin, config->wui_maps[pin].table,
+			config->wui_maps[pin].group,
+			config->wui_maps[pin].bit);
+
+	/* Determine interrupt is level or edge mode? */
+	if (mode == GPIO_INT_MODE_LEVEL)
+		miwu_mode = NPCX_MIWU_MODE_LEVEL;
+	else if (mode == GPIO_INT_MODE_EDGE)
+		miwu_mode = NPCX_MIWU_MODE_EDGE;
+
+	/* Determine trigger mode is low, high or both? */
+	if (trig == GPIO_INT_TRIG_LOW)
+		miwu_trig = NPCX_MIWU_TRIG_LOW;
+	else if (trig == GPIO_INT_TRIG_HIGH)
+		miwu_trig = NPCX_MIWU_TRIG_HIGH;
+	else if (trig == GPIO_INT_TRIG_BOTH)
+		miwu_trig = NPCX_MIWU_TRIG_BOTH;
+
+	/* Call MIWU routine to setup interrupt configuration */
+	soc_miwu_interrupt_configure(&config->wui_maps[pin],
+					miwu_mode, miwu_trig);
+
+	/* Enable/Disable irq of wake-up input sources */
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		soc_miwu_irq_disable(&config->wui_maps[pin]);
+	} else {
+		soc_miwu_irq_enable(&config->wui_maps[pin]);
+	}
+
+	return 0;
+}
+
+static int gpio_npcx_manage_callback(struct device *dev,
+				      struct gpio_callback *callback, bool set)
+{
+	const struct gpio_npcx_config *config = DRV_CONFIG(dev);
+	struct miwu_io_callback *miwu_cb = (struct miwu_io_callback *)callback;
+	int pin = find_lsb_set(callback->pin_mask) - 1;
+
+	/* Has the IO pin valid MIWU input source? */
+	if (config->wui_maps[pin].table == NPCX_MIWU_TABLE_NONE) {
```

Comment:
should we validate that pin != -1 before accessing into array? What guarantee do we have the callback->pin_mask is not 0? (If we do have those guarantees, then we should call them out in comments here)

Reply:
Done.